File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/embedded/EmbeddedTest.java
package org.babyfish.jimmer.sql.embedded;

import org.babyfish.jimmer.sql.common.AbstractQueryTest;
import org.babyfish.jimmer.sql.model.embedded.*;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;

public class EmbeddedTest extends AbstractQueryTest {

    private static final String ROWS = "[" +
            "--->{" +
            "--->--->\"id\":1," +
            "--->--->\"source\":{" +
            "--->--->--->\"leftTop\":{\"x\":100,\"y\":120}," +
            "--->--->--->\"rightBottom\":{\"x\":400,\"y\":320}" +
            "--->--->}," +
            "--->--->\"target\":{" +
            "--->--->--->\"leftTop\":{\"x\":800,\"y\":600}," +
            "--->--->--->\"rightBottom\":{\"x\":1400,\"y\":1000}" +
            "--->--->}" +
            "--->}" +
            "]";

    @Test
    public void testFindById() {
        TransformTable transform = TransformTable.$;
        executeAndExpect(
                getSqlClient()
                        .createQuery(transform)
                        .where(transform.id().eq(1L))
                        .select(transform),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, " +
                                    "tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, " +
                                    "tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM " +
                                    "from TRANSFORM as tb_1_ " +
                                    "where tb_1_.ID = ?"
                    ).variables(1L);
                    ctx.rows(ROWS);
                }
        );
    }

    @Test
    public void testFindNullById() {
        TransformTable transform = TransformTable.$;
        executeAndExpect(
                getSqlClient()
                        .createQuery(transform)
                        .where(transform.id().eq(2L))
                        .select(transform),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, " +
                                    "tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, " +
                                    "tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM " +
                                    "from TRANSFORM as tb_1_ " +
                                    "where tb_1_.ID = ?"
                    ).variables(2L);
                    ctx.rows(
                            "[" +
                                    "--->{" +
                                    "--->--->\"id\":2," +
                                    "--->--->\"source\":{" +
                                    "--->--->--->\"leftTop\":{\"x\":150,\"y\":170}," +
                                    "--->--->--->\"rightBottom\":{\"x\":450,\"y\":370}" +
                                    "--->--->}," +
                                    "--->--->\"target\":null" +
                                    "--->}" +
                                    "]"
                    );
                }
        );
    }

    @Test
    public void testFindBySourceLeftTopX() {
        TransformTable transform = TransformTable.$;
        executeAndExpect(
                getSqlClient()
                        .createQuery(transform)
                        .where(transform.source().leftTop().x().eq(100L))
                        .select(transform),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, " +
                                    "tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, " +
                                    "tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM " +
                                    "from TRANSFORM as tb_1_ " +
                                    "where tb_1_.`LEFT` = ?"
                    ).variables(100L);
                    ctx.rows(ROWS);
                }
        );
    }

    @Test
    public void testFindBySourceLeftTop() {
        TransformTable transform = TransformTable.$;
        Point point = PointDraft.$.produce(draft -> draft.setX(100).setY(120));
        executeAndExpect(
                getSqlClient()
                        .createQuery(transform)
                        .where(transform.source().leftTop().eq(point))
                        .select(transform),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, " +
                                    "tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, " +
                                    "tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM " +
                                    "from TRANSFORM as tb_1_ " +
                                    "where (tb_1_.`LEFT`, tb_1_.TOP) = (?, ?)"
                    ).variables(100L, 120L);
                    ctx.rows(ROWS);
                }
        );
    }

    @Test
    public void testFindBySource() {
        TransformTable transform = TransformTable.$;
        Rect rect = RectDraft.$.produce(draft ->
                draft
                        .setLeftTop(leftTop -> leftTop.setX(100).setY(120))
                        .setRightBottom(leftTop -> leftTop.setX(400).setY(320))
        );
        executeAndExpect(
                getSqlClient()
                        .createQuery(transform)
                        .where(transform.source().eq(rect))
                        .select(transform),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, " +
                                    "tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, " +
                                    "tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM " +
                                    "from TRANSFORM as tb_1_ " +
                                    "where (tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM) = (?, ?, ?, ?)"
                    ).variables(100L, 120L, 400L, 320L);
                    ctx.rows(ROWS);
                }
        );
    }

    @Test
    public void testSelectLevel1() {
        TransformTable transform = TransformTable.$;
        executeAndExpect(
                getSqlClient()
                        .createQuery(transform)
                        .where(transform.id().eq(1L))
                        .select(
                                transform.id(),
                                transform.source(),
                                transform.target()
                        ),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, " +
                                    "tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, " +
                                    "tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM " +
                                    "from TRANSFORM as tb_1_ " +
                                    "where tb_1_.ID = ?"
                    );
                    ctx.rows(
                            "[" +
                                    "--->{" +
                                    "--->--->\"_1\":1," +
                                    "--->--->\"_2\":{" +
                                    "--->--->--->\"leftTop\":{\"x\":100,\"y\":120}," +
                                    "--->--->--->\"rightBottom\":{\"x\":400,\"y\":320}" +
                                    "--->--->}," +
                                    "--->--->\"_3\":{" +
                                    "--->--->--->\"leftTop\":{\"x\":800,\"y\":600}," +
                                    "--->--->--->\"rightBottom\":{\"x\":1400,\"y\":1000}" +
                                    "--->--->}" +
                                    "--->}" +
                                    "]"
                    );
                }
        );
    }

    @Test
    public void testSelectLevel2() {
        TransformTable transform = TransformTable.$;
        executeAndExpect(
                getSqlClient()
                        .createQuery(transform)
                        .where(transform.id().eq(1L))
                        .select(
                                transform.id(),
                                transform.source().leftTop(),
                                transform.source().rightBottom(),
                                transform.target().leftTop(),
                                transform.target().rightBottom()
                        ),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, " +
                                    "tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, " +
                                    "tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM " +
                                    "from TRANSFORM as tb_1_ " +
                                    "where tb_1_.ID = ?"
                    );
                    ctx.rows(
                            "[" +
                                    "--->{" +
                                    "--->--->\"_1\":1," +
                                    "--->--->\"_2\":{\"x\":100,\"y\":120}," +
                                    "--->--->\"_3\":{\"x\":400,\"y\":320}," +
                                    "--->--->\"_4\":{\"x\":800,\"y\":600}," +
                                    "--->--->\"_5\":{\"x\":1400,\"y\":1000}" +
                                    "--->}" +
                                    "]"
                    );
                }
        );
    }

    @Test
    public void testSelectLevel3() {
        TransformTable transform = TransformTable.$;
        executeAndExpect(
                getSqlClient()
                        .createQuery(transform)
                        .where(transform.id().eq(1L))
                        .select(
                                transform.id(),
                                transform.source().leftTop().x(),
                                transform.source().leftTop().y(),
                                transform.source().rightBottom().x(),
                                transform.source().rightBottom().y(),
                                transform.target().leftTop().x(),
                                transform.target().leftTop().y(),
                                transform.target().rightBottom().x(),
                                transform.target().rightBottom().y()
                        ),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, " +
                                    "tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, " +
                                    "tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM " +
                                    "from TRANSFORM as tb_1_ " +
                                    "where tb_1_.ID = ?"
                    );
                    ctx.rows(
                            "[" +
                                    "--->{" +
                                    "--->--->\"_1\":1," +
                                    "--->--->\"_2\":100," +
                                    "--->--->\"_3\":120," +
                                    "--->--->\"_4\":400," +
                                    "--->--->\"_5\":320," +
                                    "--->--->\"_6\":800," +
                                    "--->--->\"_7\":600," +
                                    "--->--->\"_8\":1400," +
                                    "--->--->\"_9\":1000" +
                                    "--->}" +
                                    "]"
                    );
                }
        );
    }

    @Test
    public void findByCompositeIds() {
        List<OrderItemId> orderItemIds = Arrays.asList(
                OrderItemIdDraft.$.produce(draft -> draft.setA(1).setB(1).setC(1)),
                OrderItemIdDraft.$.produce(draft -> draft.setA(1).setB(1).setC(2))
        );
        anyAndExpect(
                con -> getSqlClient()
                        .getEntities()
                        .forConnection(con)
                        .findByIds(OrderItem.class, orderItemIds),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ORDER_ITEM_A, tb_1_.ORDER_ITEM_B, tb_1_.ORDER_ITEM_C, " +
                                    "--->tb_1_.NAME, tb_1_.FK_ORDER_X, tb_1_.FK_ORDER_Y " +
                                    "from ORDER_ITEM as tb_1_ " +
                                    "where (" +
                                    "--->tb_1_.ORDER_ITEM_A, tb_1_.ORDER_ITEM_B, tb_1_.ORDER_ITEM_C" +
                                    ") in (" +
                                    "--->(?, ?, ?), (?, ?, ?)" +
                                    ")"
                    );
                }
        );
    }

    @Test
    public void testNullCompositeId() {
        OrderItemTable orderItem = OrderItemTable.$;
        executeAndExpect(
                getSqlClient()
                        .createQuery(orderItem)
                        .where(orderItem.order().isNull())
                        .select(orderItem),
                ctx -> {
                    ctx.sql(
                            "select " +
                                    "--->tb_1_.ORDER_ITEM_A, tb_1_.ORDER_ITEM_B, tb_1_.ORDER_ITEM_C, " +
                                    "--->tb_1_.NAME, tb_1_.FK_ORDER_X, tb_1_.FK_ORDER_Y " +
                                    "from ORDER_ITEM as tb_1_ " +
                                    "where tb_1_.FK_ORDER_X is null and tb_1_.FK_ORDER_Y is null"
                    );
                }
        );
    }

    @Test
    public void testNullSubId() {
        OrderItemTable orderItem = OrderItemTable.$;
        executeAndExpect(
                getSqlClient()
                        .createQuery(orderItem)
                        .where(orderItem.order().id().x().isNull())
                        .select(orderItem),
                ctx -> {
                    ctx.sql(
                            "select " +
                                    "--->tb_1_.ORDER_ITEM_A, tb_1_.ORDER_ITEM_B, tb_1_.ORDER_ITEM_C, " +
                                    "--->tb_1_.NAME, tb_1_.FK_ORDER_X, tb_1_.FK_ORDER_Y " +
                                    "from ORDER_ITEM as tb_1_ " +
                                    "where tb_1_.FK_ORDER_X is null"
                    );
                }
        );
    }

    @Test
    public void testNonNullCompositeId() {
        OrderItemTable orderItem = OrderItemTable.$;
        executeAndExpect(
                getSqlClient()
                        .createQuery(orderItem)
                        .where(orderItem.order().isNotNull())
                        .select(orderItem),
                ctx -> {
                    ctx.sql(
                            "select " +
                                    "--->tb_1_.ORDER_ITEM_A, tb_1_.ORDER_ITEM_B, tb_1_.ORDER_ITEM_C, " +
                                    "--->tb_1_.NAME, tb_1_.FK_ORDER_X, tb_1_.FK_ORDER_Y " +
                                    "from ORDER_ITEM as tb_1_ " +
                                    "where tb_1_.FK_ORDER_X is not null and tb_1_.FK_ORDER_Y is not null"
                    );
                }
        );
    }

    @Test
    public void testNonNullSubId() {
        OrderItemTable orderItem = OrderItemTable.$;
        executeAndExpect(
                getSqlClient()
                        .createQuery(orderItem)
                        .where(orderItem.order().id().x().isNotNull())
                        .select(orderItem),
                ctx -> {
                    ctx.sql(
                            "select " +
                                    "--->tb_1_.ORDER_ITEM_A, tb_1_.ORDER_ITEM_B, tb_1_.ORDER_ITEM_C, " +
                                    "--->tb_1_.NAME, tb_1_.FK_ORDER_X, tb_1_.FK_ORDER_Y " +
                                    "from ORDER_ITEM as tb_1_ " +
                                    "where tb_1_.FK_ORDER_X is not null"
                    );
                }
        );
    }
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/query/EmbeddedTest.kt
package org.babyfish.jimmer.sql.kt.query

import org.babyfish.jimmer.kt.new
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.babyfish.jimmer.sql.kt.ast.expression.eq
import org.babyfish.jimmer.sql.kt.common.AbstractQueryTest
import org.babyfish.jimmer.sql.kt.common.assertContentEquals
import org.babyfish.jimmer.sql.kt.model.embedded.*
import org.babyfish.jimmer.sql.kt.model.embedded.dto.TransformFlatView
import org.babyfish.jimmer.sql.kt.model.embedded.dto.TransformSpecification
import org.babyfish.jimmer.sql.kt.model.embedded.dto.TransformView
import org.babyfish.jimmer.sql.kt.model.embedded.dto.TransformView2
import kotlin.test.Test

class EmbeddedTest : AbstractQueryTest() {

    @Test
    fun testFindBySourceLeftTopX() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                where(table.source.leftTop.x eq 100)
                select(table)
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, 
                    |tb_1_.BOTTOM, tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM 
                    |from TRANSFORM tb_1_ 
                    |where tb_1_.`LEFT` = ?""".trimMargin()
            )
            variables(100L)
            rows(ROWS)
        }
    }

    @Test
    fun testFindBySourceLeftTop() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                where(table.source.leftTop eq new(Point::class).by {
                    x = 100
                    y = 120
                })
                select(table)
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, 
                    |tb_1_.BOTTOM, tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM 
                    |from TRANSFORM tb_1_ 
                    |where (tb_1_.`LEFT`, tb_1_.TOP) = (?, ?)""".trimMargin()
            )
            variables(100L, 120L)
            rows(ROWS)
        }
    }

    @Test
    fun testFindBySource() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                where(table.source eq new(Rect::class).by {
                    leftTop().apply {
                        x = 100
                        y = 120
                    }
                    rightBottom().apply {
                        x = 400
                        y = 320
                    }
                })
                select(table)
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, 
                    |tb_1_.BOTTOM, tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM 
                    |from TRANSFORM tb_1_ 
                    |where (tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM) = (?, ?, ?, ?)""".trimMargin()
            )
            variables(100L, 120L, 400L, 320L)
            rows(ROWS)
        }
    }

    @Test
    fun testSelectLevel1() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                select(
                    table.id,
                    table.source,
                    table.target
                )
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, 
                    |tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM 
                    |from TRANSFORM tb_1_""".trimMargin()
            )
            rows(
                """[
                    |--->{
                    |--->--->"_1":1,
                    |--->--->"_2":{
                    |--->--->--->"leftTop":{"x":100,"y":120},
                    |--->--->--->"rightBottom":{"x":400,"y":320}
                    |--->--->},
                    |--->--->"_3":{
                    |--->--->--->"leftTop":{"x":800,"y":600},
                    |--->--->--->"rightBottom":{"x":1400,"y":1000}
                    |--->--->}
                    |--->}
                    |]""".trimMargin()
            )
        }
    }

    @Test
    fun testSelectLevel2() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                select(
                    table.id,
                    table.source.leftTop,
                    table.source.rightBottom,
                    table.target.leftTop,
                    table.target.rightBottom
                )
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, 
                    |tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM 
                    |from TRANSFORM tb_1_""".trimMargin()
            )
            rows(
                """[
                    |--->{
                    |--->--->"_1":1,
                    |--->--->"_2":{"x":100,"y":120},
                    |--->--->"_3":{"x":400,"y":320},
                    |--->--->"_4":{"x":800,"y":600},
                    |--->--->"_5":{"x":1400,"y":1000}
                    |--->}
                    |]""".trimMargin()
            )
        }
    }

    @Test
    fun testSelectLevel3() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                select(
                    table.id,
                    table.source.leftTop.x,
                    table.source.leftTop.y,
                    table.source.rightBottom.x,
                    table.source.rightBottom.y,
                    table.target.leftTop.x,
                    table.target.leftTop.y,
                    table.target.rightBottom.x,
                    table.target.rightBottom.y
                )
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, 
                    |tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM 
                    |from TRANSFORM tb_1_""".trimMargin()
            )
            rows(
                """[
                    |--->{
                    |--->--->"_1":1,
                    |--->--->"_2":100,
                    |--->--->"_3":120,
                    |--->--->"_4":400,
                    |--->--->"_5":320,
                    |--->--->"_6":800,
                    |--->--->"_7":600,
                    |--->--->"_8":1400,
                    |--->--->"_9":1000
                    |--->}
                    |]""".trimMargin()
            )
        }
    }

    @Test
    fun testObjectFetcher() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                select(
                    table.fetchBy {
                        source {
                            leftTop {
                                x()
                            }
                        }
                        target {
                            rightBottom {
                                y()
                            }
                        }
                    }
                )
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.`LEFT`, tb_1_.TARGET_BOTTOM 
                    |from TRANSFORM tb_1_""".trimMargin()
            )
            rows(
                """[{"id":1,"source":{"leftTop":{"x":100}},"target":{"rightBottom":{"y":1000}}}]"""
            )
        }
    }

    @Test
    fun testDto() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                select(
                    table.fetch(TransformView::class)
                )
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.`LEFT`, tb_1_.TARGET_BOTTOM 
                    |from TRANSFORM tb_1_""".trimMargin()
            )
            rows {
                assertContentEquals(
                    """[
                        |--->TransformView(
                        |--->--->id=1, 
                        |--->--->source=TransformView.TargetOf_source(
                        |--->--->--->leftTop=TransformView.TargetOf_source.TargetOf_leftTop(x=100)
                        |--->--->), 
                        |--->--->target=TransformView.TargetOf_target(
                        |--->--->--->rightBottom=TransformView.TargetOf_target.TargetOf_rightBottom(
                        |--->--->--->--->y=1000
                        |--->--->--->)
                        |--->--->)
                        |--->)
                        |]""".trimMargin(),
                    it
                )
            }
        }
    }

    @Test
    fun testDtoWithFormula() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                select(
                    table.fetch(TransformView2::class)
                )
            }
        ) {
            sql(
                """select tb_1_.ID, 
                    |tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, 
                    |tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM, tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP 
                    |from TRANSFORM tb_1_""".trimMargin()
            )
            rows {
                assertContentEquals(
                    "[" +
                        "--->TransformView2(" +
                        "--->--->id=1, " +
                        "--->--->source=TransformView2.TargetOf_source(" +
                        "--->--->--->area=60000, " +
                        "--->--->--->leftTop=TransformView2.TargetOf_source.TargetOf_leftTop(x=100)" +
                        "--->--->), " +
                        "--->--->target=TransformView2.TargetOf_target(" +
                        "--->--->--->area=240000, " +
                        "--->--->--->rightBottom=TransformView2.TargetOf_target.TargetOf_rightBottom(y=1000)" +
                        "--->--->)" +
                        "--->)" +
                        "]",
                    it
                )
            }
        }
    }

    @Test
    fun testFlatDto() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                select(
                    table.fetch(TransformFlatView::class)
                )
            }
        ) {
            sql(
                """select tb_1_.ID, 
                    |tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, 
                    |tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM 
                    |from TRANSFORM tb_1_""".trimMargin()
            )
            rows {
                assertContentEquals(
                    """[
                        |--->TransformFlatView(
                        |--->--->id=1, 
                        |--->--->sourceX1=100, 
                        |--->--->sourceY1=120, 
                        |--->--->sourceX2=400, 
                        |--->--->sourceY2=320, 
                        |--->--->targetX1=800, 
                        |--->--->targetY1=600, 
                        |--->--->targetX2=1400, 
                        |--->--->targetY2=1000
                        |--->)
                        |]""".trimMargin(),
                    it
                )
            }
        }
    }

    @Test
    fun testFormulaDependsOnEmbeddable() {
        executeAndExpect(
            sqlClient.createQuery(Machine::class) {
                select(
                    table.fetchBy {
                        factoryCount()
                        factoryNames()
                        detail {
                            patents()
                        }
                    }
                )
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.patent_map, tb_1_.factory_map from MACHINE tb_1_"""
            )
            rows(
                """[
                    |--->{
                    |--->--->"id":1,
                    |--->--->"detail":{"patents":{"p-1":"patent-1","p-2":"patent-2"}},
                    |--->--->"factoryCount":2,
                    |--->--->"factoryNames":["f-1","f-2"]
                    |--->}
                    |]""".trimMargin()
            )
        }
    }

    @Test
    fun testFormulaDependsOnDuplicatedEmbeddable() {
        executeAndExpect(
            sqlClient.createQuery(Machine::class) {
                select(
                    table.fetchBy {
                        factoryCount()
                        detail {
                            patents()
                            factories()
                        }
                    }
                )
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.patent_map, tb_1_.factory_map from MACHINE tb_1_"""
            )
            rows(
                """[
                    |--->{
                    |--->--->"id":1,
                    |--->--->"detail":{
                    |--->--->--->"factories":{"f-1":"factory-1","f-2":"factory-2"},
                    |--->--->--->"patents":{"p-1":"patent-1","p-2":"patent-2"}
                    |--->--->},
                    |--->--->"factoryCount":2
                    |--->}
                    |]""".trimMargin()
            )
        }
    }

    @Test
    fun testFormulaInEmbeddable() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                select(
                    table.fetchBy {
                        source {
                            area()
                        }
                        target {
                            area()
                        }
                    }
                )
            }
        ) {
            sql(
                """select tb_1_.ID, 
                    |tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, 
                    |tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM 
                    |from TRANSFORM tb_1_""".trimMargin()
            )
            rows(
                """[{"id":1,"source":{"area":60000},"target":{"area":240000}}]"""
            )
        }
    }

    @Test
    fun testFormulaInEmbeddableWithDuplicatedFetching() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                select(
                    table.fetchBy {
                        source {
                            area()
                            leftTop {
                                x()
                            }
                        }
                        target {
                            area()
                            rightBottom {
                                y()
                            }
                        }
                    }
                )
            }
        ) {
            sql(
                """select tb_1_.ID, 
                    |tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, 
                    |tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM, tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP 
                    |from TRANSFORM tb_1_""".trimMargin()
            )
            rows(
                """[
                    |--->{
                    |--->--->"id":1,
                    |--->--->"source":{
                    |--->--->--->"leftTop":{"x":100},
                    |--->--->--->"area":60000
                    |--->--->},
                    |--->--->"target":{
                    |--->--->--->"rightBottom":{"y":1000},
                    |--->--->--->"area":240000
                    |--->--->}
                    |--->}
                    |]""".trimMargin()
            )
        }
    }

    companion object {
        private val ROWS = """[
            |--->{
            |--->--->"id":1,
            |--->--->"source":{
            |--->--->--->"leftTop":{"x":100,"y":120},
            |--->--->--->"rightBottom":{"x":400,"y":320}
            |--->--->},
            |--->--->"target":{
            |--->--->--->"leftTop":{"x":800,"y":600},
            |--->--->--->"rightBottom":{"x":1400,"y":1000}
            |--->--->}
            |--->}
            |]""".trimMargin()
    }

    @Test
    fun testSpecification() {
        executeAndExpect(
            sqlClient.createQuery(Transform::class) {
                where(
                    TransformSpecification(
                        minX = 100L,
                        maxX = 2000L
                    )
                )
                select(table)
            }
        ) {
            sql(
                """select tb_1_.ID, 
                    |tb_1_.`LEFT`, tb_1_.TOP, tb_1_.`RIGHT`, tb_1_.BOTTOM, 
                    |tb_1_.TARGET_LEFT, tb_1_.TARGET_TOP, tb_1_.TARGET_RIGHT, tb_1_.TARGET_BOTTOM 
                    |from TRANSFORM tb_1_ 
                    |where tb_1_.`LEFT` >= ? and tb_1_.TARGET_RIGHT <= ?""".trimMargin()
            )
            rows(
                """[{
                    |--->"id":1,
                    |--->"source":{
                    |--->--->"leftTop":{"x":100,"y":120},
                    |--->--->"rightBottom":{"x":400,"y":320}
                    |--->},
                    |--->"target":{
                    |--->--->"leftTop":{"x":800,"y":600},
                    |--->--->"rightBottom":{"x":1400,"y":1000}
                    |--->}
                    |}]""".trimMargin()
            )
        }
    }
}